server.go

```diff
@@ -296,25 +304,79 @@ type Server struct {
 	// DecorateWriter is optional, allows customization of the process that writes raw DNS messages.
 	DecorateWriter DecorateWriter
 
+	// UDP packet or TCP connection queue
+	queue chan *response
+	// Workers count
+	workersCount int32
 	// Shutdown handling
 	lock    sync.RWMutex
 	started bool
 }
 
+func (srv *Server) worker(w *response) {
+	defer atomic.AddInt32(&srv.workersCount, -1)
+
+	srv.serve(w)
+
+	count := 0
+	ticker := time.NewTicker(idleWorkerTimeout)
+	defer ticker.Stop()
+LOOP:
+	for {
+		select {
+		case w, ok := <-srv.queue:
+			if !ok {
+				break LOOP
+			}
+			count += 1
+			srv.serve(w)
+		case <-ticker.C:
+			if count == 0 {
+				break LOOP
+			}
+			count = 0
+		}
+	}
+}
+
+func (srv *Server) spawnWorker(w *response) {
+	select {
+	case srv.queue <- w:
+		return
+	default:
+	}
+
+	for {
```

Comment:
This doesn’t work as well here and dropping requests on the floor is very much the wrong approach. It leads to trivial DOS attacks against the TCP server in particular which would be made worse by removing `maxTCPQueries`.

Either move this back into `worker` with a `serveDNS` before `w.Close` or replace the `w.Close` below with a blocking send.

Reply:
If we move check to worker() - it means that maxWorkersCount makes no sence.
10000 workers are busy and we continue spawn goroutines.
We cannot prevent DOS in any case - the difference is that your proposal doesn't prevent out of resources - but current code does.

Reply:
This approach has a trivial DOS vector, the existing code doesn’t which is the key point to me. The existing code, and my suggestion to move the check, will scale as far as the host machine will. This code has an artificial hard limit of 10000, which is arbitrary. There are systems which can scale farther than that. The `maxWorkersCount` would still have value because it would limit the long-lived goroutines.

Some quick numbers: `_StackMin` is 2KiB meaning 10,000 goroutines uses `2048*10000/2^20 = 19GiB` of memory for stack. That’s too large for many systems and far too small for many others.

I like the approach of the check being in the goroutine because it saves the overhead of creating and destroying goroutines, but doesn’t limit the servers maximum performance. This one just drops requests on the floor which feels very wrong to me.

I think conflating resource control with long-lived worker goroutines is a mistake, because they’re too distinct issues.

Reply:
Done

Reply:
BTW 2KiB * 10000 ~ 20 MiB not 20 GiB

Reply:
Whoops @ my math.